### ISSUE TITLE
groupby on a filtered frame should automatically persist

### ISSUE THREAD

**Issue by jreback** [dask/distributed#617]
so when doing a filtered frame (not using the index, maybe that is a different case), then doing an intensive operation, like a groupby. It is optimal to ``.persist`` the intermediate before the final computation to avoid many recompuations (e.g. for each of the groupers at the very least).

![image](https://cloud.githubusercontent.com/assets/953992/19929017/a34cf8f2-a0d7-11e6-950b-3f6a85dc7ecb.png)

![image](https://cloud.githubusercontent.com/assets/953992/19929023/ab2a539e-a0d7-11e6-8dee-274e5eae926d.png)

**Comment by mrocklin** [comment]
The opportunistic caching logic that backs [cachey](https://github.com/dask/cachey) ([blogpost](http://matthewrocklin.com/blog/work/2015/08/03/Caching)) would be a good fit here.  It would tell us which tasks we should hold in a finite amount of memory.

I think that we could integrate this into the scheduler fairly easily by using a [SchedulerPlugin](http://distributed.readthedocs.io/en/latest/plugins.html) to watch all computations as they happen (which gives us the runtime and storage costs of all results) and maintain a fake client to hold on to keys that we find valuable.  I think that the scheduler's existing system can manage the rest of the logic.

**Comment by martindurant** [comment]
I did like the simple-but-powerful heuristic of cachey.
Implementing a fake client to hold the references could be similar to the publish_dataset stuff, but we'd have to think a little about when to call cachey (or any arbitrary caching mechanism - could be LRU) in the flow of a task.

**Comment by kynan** [comment]
I have been thinking about how to best implement an automatic caching of intermediate results.

Using a `SchedulerPlugin`as @mrocklin suggests I assume the place to hook in would be the transition processing -> memory since this is where we have all the information to make a decision about whether to keep a result (or pass the information to e.g. cachey). Hooking in to memory -> released is too late afaiu since the plugins are called *after* the transition has finished and can't influence the outcome.

Rather than having a global cap on the amount of data cached I think it may be preferable to have a separate cache per worker.

Does the following implementation strategy seem sensible?
* have a SchedulerPlugin with a `transition` hook
* have a cachey Cache as "fake client" per worker with a configurable max size
* on a processing -> memory transition, put the key in the cache of the that worker with compute time as cost
* on a memory -> released / forgotten transition, purge the key from the cache
* subclass cachey.Cache to add hooks for calling `client_wants_keys` when a new key is added and `client_releases_keys` when a key is evicted

**Comment by mrocklin** [comment]
Yes, I think that processing -> memory is the right time to make the decision.

I think that it will be easier to implement if we just add a client and don't deal with caches per worker.  I don't think that we should use the actual cachey project here, merely some of the logic.

You'll never get memory -> released transitions for as long as your client holds onto the value.  This is the point.  Even when no client wants a task we still hold onto it because we think that another client will come along soon that will want the same value again.

**Comment by kynan** [comment]
Would we still get memory -> forgotten transitions e.g. if a worker died?

Yes, it's probably easier to not use cachey directly. In fact, could we not use the worker's memory limit for deciding when to evict?

**Comment by mrocklin** [comment]
I would recommend that caching at the cluster level not think at all about individual workers and instead we trust the scheduler to move data around if necessary or we trust workers to store data to disk if necessary.

The forgotten state means that the scheduler is choosing to lose all information about the state.  If a worker dies then you might get memory -> released, but not forgotten.

**Comment by kynan** [comment]
Seems when inserting keys as a "fake client" (named `scheduler-cache`) we can actually use the scheduler's internal bookkeeping and don't need to manually keep track of what we have cached.

Deciding what to cache however turn out to be non-trivial and I think a cost function alone is not sufficient. I don't think we ever want to cache e.g. intermediate steps of an aggregation like `mean`.

As a first heuristic (which is probably too simplistic) I'm only considering keys another client (other than `scheduler-cache`) has requested for caching. Is that a reasonable assumption to make or is it likely we might want to keep intermediate results for other kinds of computations around?

**Comment by mrocklin** [comment]
You should experiment with whatever you think is likely to be valuable.  I
have my own thoughts on this topic, but they are by no means the right way
to do things.

On Tue, Nov 29, 2016 at 1:19 PM, Florian Rathgeber <notifications@github.com
